### ISSUE TITLE
Listen to channel state, using the singleton completion queue, to uncache them when they fail

### ISSUE THREAD

**Issue by jcanizales** [grpc/grpc#5851]
If we don't explicitly listen to channel state, we'll only known of failed channels (e.g. ones that received a `goaway` when no RPC was pending) once the next RPC is attempted (and fails). This is intended behavior of the C layer, for the datacenter use case. Mobile wrappers are thus expected to do this to attain the behavior that's preferable on mobile devices.

**Comment by makdharma** [comment]
@jcanizales Can you explain in a bit more detail what needs to be done here? Is this still applicable after your network status change?

**Comment by jcanizales** [comment]
The C library has an API to listen to changes in channel state (connecting, ready, transient failure, fatal failure, ...). We need to subscribe to those notifications in addition to the ones we already have from the OS. The reason is that sometimes a server closes a connection (by sending `goaway`) without there being any connectivity problems (e.g. after some timeout of it being idle). If we don't listen to channel state, the C library will be oblivious of it, the `goaway` will just remain enqueued, and the next RPC to that host will invariably fail. By subscribing, the C library does look at those things.

**Comment by makdharma** [comment]
Makes sense. Do we have (or need) an interop test that sends GOAWAY from server?

**Comment by jcanizales** [comment]
We don't have one that I know of, and it's a great idea. @yang-g would you add that to the C++ interop server?

**Comment by makdharma** [comment]
I wrote a test to create the scenario described in this bug without success. Looking further, I noticed that the completion queue in ObjectiveC is constantly polled in a while loop. This makes C core notice the goaway and connection issues without needing to spin another completion queue or starting a watcher. See here https://github.com/grpc/grpc/blob/master/src/objective-c/GRPCClient/private/GRPCCompletionQueue.m#L69
This bug is actually not a valid bug. I propose closing. @jcanizales , @ctiller - any objections?
